### PR TITLE
DROOLS-3930: Improve usability from Guided Rule Template - Can't resize column

### PR DESCRIPTION
**Pull request by manstis**
See https://issues.jboss.org/browse/DROOLS-3930

Unfortunately _lots_of formatting changes; but I'll call out the main changes below.

### CODE REVIEW

File: kie-wb-common-widgets/kie-wb-decorated-grid-widget/src/main/java/org/kie/workbench/common/widgets/decoratedgrid/client/widget/AbstractDecoratedGridHeaderWidget.java [hunk 1]
```diff
@@ -82,27 +87,54 @@
         private DynamicColumn<T> resizeColumn = null;
         private int resizeColumnWidth = 0;
 
+        public boolean isResizePrimed() {
```
Comment (manstis) [inline kie-wb-common-widgets/kie-wb-decorated-grid-widget/src/main/java/org/kie/workbench/common/widgets/decoratedgrid/client/widget/AbstractDecoratedGridHeaderWidget.java]: Some more _getters_ and _setters_ to an _inner_ class to help with Unit Testing.

File: kie-wb-common-widgets/kie-wb-decorated-grid-widget/src/main/java/org/kie/workbench/common/widgets/decoratedgrid/client/widget/AbstractDecoratedGridHeaderWidget.java [hunk 2]
```diff
@@ -129,141 +161,157 @@ public void setResizePrimed( boolean isResizePrimed ) {
...
-        Element tre = DOM.createTR();
-        Element tce = DOM.createTD();
-        Element div = DOM.createDiv();
-        div.getStyle().setPosition( Position.RELATIVE );
-        getBody().getParentElement().<TableElement>cast().setCellSpacing( 0 );
-        getBody().getParentElement().<TableElement>cast().setCellPadding( 0 );
-        tce.appendChild( div );
-        tre.appendChild( tce );
-        getBody().appendChild( tre );
+        TableRowElement tre = Document.get().createTRElement();
```
Comment (manstis) [inline kie-wb-common-widgets/kie-wb-decorated-grid-widget/src/main/java/org/kie/workbench/common/widgets/decoratedgrid/client/widget/AbstractDecoratedGridHeaderWidget.java]: Tweaks to how the `DOM` elements are created for Unit Tests.

File: kie-wb-common-widgets/kie-wb-decorated-grid-widget/src/main/java/org/kie/workbench/common/widgets/decoratedgrid/client/widget/AbstractDecoratedGridHeaderWidget.java [hunk 3]
```diff
@@ -129,141 +161,157 @@ public void setResizePrimed( boolean isResizePrimed ) {
...
-        resizer.getStyle().setTop( 0,
-                                   Unit.PX );
+        resizer = Document.get().createDivElement();
+        resizer.addClassName(resources.headerResizer());
+        resizer.getStyle().setTop(0,
+                                  Unit.PX);
 
         // Add the resizer to the outer most container, otherwise it gets
         // truncated by the ScrollPanel as it hides any overflow
-        div.appendChild( resizer );
```
Comment (manstis) [inline kie-wb-common-widgets/kie-wb-decorated-grid-widget/src/main/java/org/kie/workbench/common/widgets/decoratedgrid/client/widget/AbstractDecoratedGridHeaderWidget.java]: Don't attach the "column resize handlers" to the _header_ widget........

File: kie-wb-common-widgets/kie-wb-decorated-grid-widget/src/main/java/org/kie/workbench/common/widgets/decoratedgrid/client/widget/AbstractDecoratedGridHeaderWidget.java [hunk 4]
```diff
@@ -129,141 +161,157 @@ public void setResizePrimed( boolean isResizePrimed ) {
...
+        eventBus.addHandler(SetColumnVisibilityEvent.TYPE,
+                            this);
+        eventBus.addHandler(UpdateColumnDefinitionEvent.TYPE,
+                            this);
+        eventBus.addHandler(MoveColumnsEvent.TYPE,
+                            this);
+    }
+
+    @Override
+    protected void onLoad() {
```
Comment (manstis) [inline kie-wb-common-widgets/kie-wb-decorated-grid-widget/src/main/java/org/kie/workbench/common/widgets/decoratedgrid/client/widget/AbstractDecoratedGridHeaderWidget.java]: ... but attach them to the `RootPanel` (whole `Document`) when this `widget` is attached to the `DOM`.

File: kie-wb-common-widgets/kie-wb-decorated-grid-widget/src/main/java/org/kie/workbench/common/widgets/decoratedgrid/client/widget/AbstractDecoratedGridHeaderWidget.java [hunk 5]
```diff
@@ -129,141 +161,157 @@ public void setResizePrimed( boolean isResizePrimed ) {
...
+
+    RootPanel rootPanel() {
+        return RootPanel.get();
     }
 
-    public HandlerRegistration addResizeHandler( ResizeHandler handler ) {
-        if ( handler == null ) {
-            throw new IllegalArgumentException( "handler cannot be null" );
+    @Override
+    protected void onUnload() {
```
Comment (manstis) [inline kie-wb-common-widgets/kie-wb-decorated-grid-widget/src/main/java/org/kie/workbench/common/widgets/decoratedgrid/client/widget/AbstractDecoratedGridHeaderWidget.java]: Make sure the "column resize handlers" are released when this `Widget` is removed from the `DOM`.

File: kie-wb-common-widgets/kie-wb-decorated-grid-widget/src/main/java/org/kie/workbench/common/widgets/decoratedgrid/client/widget/AbstractDecoratedGridHeaderWidget.java [hunk 6]
```diff
@@ -129,141 +161,157 @@ public void setResizePrimed( boolean isResizePrimed ) {
...
+                                                                                              }
+                                                                                              resizerInfo.setResizing(false);
+                                                                                              resizerInfo.setResizePrimed(false);
+                                                                                              resizer.getStyle().setVisibility(Visibility.HIDDEN);
+                                                                                              event.preventDefault();
+                                                                                          },
+                                                                                          MouseOutEvent.getType()));
+    }
+
+    RootPanel rootPanel() {
```
Comment (manstis) [inline kie-wb-common-widgets/kie-wb-decorated-grid-widget/src/main/java/org/kie/workbench/common/widgets/decoratedgrid/client/widget/AbstractDecoratedGridHeaderWidget.java]: Allow the `RootPanel` to be mocked in Unit Tests.

File: kie-wb-common-widgets/kie-wb-decorated-grid-widget/src/main/java/org/kie/workbench/common/widgets/decoratedgrid/client/widget/AbstractDecoratedGridHeaderWidget.java [hunk 7]
```diff
@@ -276,29 +324,29 @@ public HandlerRegistration addResizeHandler( ResizeHandler handler ) {
...
     // by the resizer.
-    private void setResizerDimensions( final int position ) {
-        resizer.getStyle().setHeight( parent.getElement().getClientHeight(),
-                                      Unit.PX );
-        resizer.getStyle().setLeft( position,
-                                    Unit.PX );
+    void setResizerDimensions(final int position) {
+        resizer.getStyle().setHeight(parent.getElement().getClientHeight(),
+                                     Unit.PX);
+        resizer.getStyle().setLeft(position - panel.getAbsoluteLeft(),
```
Comment (manstis) [inline kie-wb-common-widgets/kie-wb-decorated-grid-widget/src/main/java/org/kie/workbench/common/widgets/decoratedgrid/client/widget/AbstractDecoratedGridHeaderWidget.java]: Tweak for x-coordinate calculation now we're adding the event handlers to a parent panel.

File: kie-wb-common-widgets/kie-wb-decorated-grid-widget/src/test/java/org/kie/workbench/common/widgets/decoratedgrid/client/widget/AbstractDecoratedGridHeaderWidgetColumnResizeTest.java [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+import static org.junit.Assert.assertNotNull;
+import static org.mockito.Matchers.any;
+import static org.mockito.Matchers.eq;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+@RunWith(GwtMockitoTestRunner.class)
+public class AbstractDecoratedGridHeaderWidgetColumnResizeTest {
```
Comment (manstis) [inline kie-wb-common-widgets/kie-wb-decorated-grid-widget/src/test/java/org/kie/workbench/common/widgets/decoratedgrid/client/widget/AbstractDecoratedGridHeaderWidgetColumnResizeTest.java]: Maybe not exhaustive but some Unit Tests (for the "happy path") are better than none!

**Comment by jomarko** [comment]
jenkins please retest this

**Comment by jomarko** [comment]
### Manual review notes
Works fine from my point of view, just one question, column width is not preserved even between clicking different tabs (Model, Source, ...) Is it expected? 

I will proceed to code review.

File: kie-wb-common-widgets/kie-wb-decorated-grid-widget/src/main/java/org/kie/workbench/common/widgets/decoratedgrid/client/widget/AbstractDecoratedGridHeaderWidget.java [hunk 9]
```diff
@@ -129,141 +161,157 @@ public void setResizePrimed( boolean isResizePrimed ) {
...
+
+        this.rootPanelMouseOutHandlerRegistration = Optional.of(rootPanel().addDomHandler((event) -> {
+                                                                                              if (!resizerInfo.isResizing()) {
+                                                                                                  return;
+                                                                                              }
+                                                                                              resizerInfo.setResizing(false);
+                                                                                              resizerInfo.setResizePrimed(false);
+                                                                                              resizer.getStyle().setVisibility(Visibility.HIDDEN);
+                                                                                              event.preventDefault();
+                                                                                          },
```
Comment (jomarko) [inline kie-wb-common-widgets/kie-wb-decorated-grid-widget/src/main/java/org/kie/workbench/common/widgets/decoratedgrid/client/widget/AbstractDecoratedGridHeaderWidget.java]: I know you just moved an existing code, however I think we should not duplicate the code:
```
(event) -> {
    if (!resizerInfo.isResizing()) {
        return;
    }
    resizerInfo.setResizing(false);
    resizerInfo.setResizePrimed(false);
    resizer.getStyle().setVisibility(Visibility.HIDDEN);
    event.preventDefault();
}
```

Comment (manstis) [reply kie-wb-common-widgets/kie-wb-decorated-grid-widget/src/main/java/org/kie/workbench/common/widgets/decoratedgrid/client/widget/AbstractDecoratedGridHeaderWidget.java]: Updated in new commit.

**Comment by manstis** [comment]
@jomarko The editor has never preserved column width between switching tabs.. nobody has asked for it so I am reluctant to add such support to this PR. Persisting the column width requires even more extensive changes and to be honest unless a request comes through as a RFE I'd be reluctant to consider it (on a deprecated editor that really has no support other than customer tickets).

**Pull request merged**